src/fv/share/tdyugrid.c

```diff
@@ -0,0 +1,1209 @@
+#endif
+
+  PetscFunctionReturn(0);
+}
+
+/// Use the dual matrix to partition the grid
+///
+/// @param [in] DualMat A dual matrix of the adjaceny matrix
+/// @param [out] NewCellRankIS IS that identifies the rank that owns the grid cell
+/// @param [out] NewNumCellsLocal Number of local cells after partitioning
+///
+/// @returns 0 on success, or a non-zero error code on failure
+PetscErrorCode PartitionGrid(Mat DualMat, IS *NewCellRankIS, PetscInt *NewNumCellsLocal) {
+
+   PetscErrorCode ierr;
+
+  // Create partitioning object based on the dual matrix
+  MatPartitioning Part;
+  ierr = MatPartitioningCreate(PETSC_COMM_WORLD, &Part); CHKERRQ(ierr);
+  ierr = MatPartitioningSetAdjacency(Part, DualMat);CHKERRQ(ierr);
+  ierr = MatPartitioningSetFromOptions(Part);CHKERRQ(ierr);
+
+  // Now partition the mesh
+  ierr = MatPartitioningApply(Part, NewCellRankIS); CHKERRQ(ierr);
+  ierr = MatPartitioningDestroy(&Part); CHKERRQ(ierr);
+
+  // Compute the number of local grid cells on each processor
+  PetscInt commsize, myrank;
+  ierr = MPI_Comm_size(PETSC_COMM_WORLD, &commsize); CHKERRQ(ierr);
+  ierr = MPI_Comm_rank(PETSC_COMM_WORLD, &myrank); CHKERRQ(ierr);
+
+  PetscInt cell_counts[commsize];
+  ierr = ISPartitioningCount(*NewCellRankIS, commsize, cell_counts); CHKERRQ(ierr);
+  *NewNumCellsLocal = cell_counts[myrank];
+
+#ifdef UGRID_DEBUG
+  ierr = TDySavePetscISAsASCII(*NewCellRankIS,"is.out");
+#endif
+
+  PetscFunctionReturn(0);
+}
+
+/// Determine the max number of duals/neighbors of a grid cell in the mesh
+///
+/// @param [in] DualMat A dual matrix of the adjaceny matrix
+/// @param [inout] ugrid A TDyUGrid struct
+///
+/// @returns 0 on success, or a non-zero error code on failure
+static PetscErrorCode DetermineMaxNumDualCells(Mat DualMat, TDyUGrid *ugrid) {
+
+  PetscErrorCode ierr;
+
+  const PetscInt *ia_ptr, *ja_ptr;
+  PetscBool success;
+  PetscInt num_rows;
+  ierr = MatGetRowIJ(DualMat, 0, PETSC_FALSE, PETSC_FALSE, &num_rows, &ia_ptr, &ja_ptr, &success);CHKERRQ(ierr);
+
+  if (!success) {
+    MPI_Comm comm;
+    ierr = PetscObjectGetComm((PetscObject)DualMat, &comm); CHKERRQ(ierr);
+    SETERRQ(comm, PETSC_ERR_USER, "Error get row and column indices from dual matrix");
+  }
+
+  ugrid->max_ndual_per_cell = 0;
+  for (PetscInt icell=0; icell<ugrid->num_cells_local; icell++) {
+    PetscInt istart = ia_ptr[icell];
+    PetscInt iend = ia_ptr[icell+1];
+    PetscInt num_cols = iend - istart;
+    if (num_cols > ugrid->max_ndual_per_cell) {
+      ugrid->max_ndual_per_cell = num_cols;
+    }
+  }
+
+  PetscInt tmp = ugrid->max_ndual_per_cell;
+  ierr = MPI_Allreduce(&tmp, &ugrid->max_ndual_per_cell, 1, MPI_INTEGER, MPI_MAX, PETSC_COMM_WORLD);
+
+  ierr = MatRestoreRowIJ(DualMat, 0, PETSC_FALSE, PETSC_FALSE, &num_rows, &ia_ptr, &ja_ptr, &success);CHKERRQ(ierr);
+
+  PetscFunctionReturn(0);
+
+}
+
+/// Create IS the maps a natural-ordered Vec to a PETSC-ordered Vec
```

Comment:
Is "PETSc order" the same as a "global ordering"? If so, let's use that terminology.